contracts/pools/idle/StableSwapIDLE.vy

```diff
@@ -0,0 +1,998 @@
+        if amount != 0:
+            coin: address = self.coins[i]
+            if _use_underlying:
+                amount = IdleToken(coin).redeemIdleToken(amount)
+                ERC20(self.underlying_coins[i]).transfer(msg.sender, amount)
+            else:
+                ERC20(coin).transfer(msg.sender, amount)
+
+    log RemoveLiquidityImbalance(msg.sender, amounts, fees, D1, token_supply - token_amount)
+
+    return token_amount
+
+
+@pure
+@internal
+def get_y_D(A_: uint256, i: int128, xp: uint256[N_COINS], D: uint256) -> uint256:
+    """
+    Calculate x[i] if one reduces D from being calculated for xp to D
+
+    Done by solving quadratic equation iteratively.
+    x_1**2 + x1 * (sum' - (A*n**n - 1) * D / (A * n**n)) = D ** (n + 1) / (n ** (2 * n) * prod' * A)
+    x_1**2 + b*x_1 = c
+
+    x_1 = (x_1**2 + c) / (2*x_1 + b)
+    """
+    # x in the input is converted to the same price/precision
+
+    assert i >= 0  # dev: i below zero
+    assert i < N_COINS  # dev: i above N_COINS
+
+    Ann: uint256 = A_ * N_COINS
+    c: uint256 = D
+    S_: uint256 = 0
+    _x: uint256 = 0
+    y_prev: uint256 = 0
+
+    for _i in range(N_COINS):
+        if _i != i:
+            _x = xp[_i]
+        else:
+            continue
+        S_ += _x
+        c = c * D / (_x * N_COINS)
+    c = c * D * A_PRECISION / (Ann * N_COINS)
+    b: uint256 = S_ + D * A_PRECISION / Ann
+    y: uint256 = D
+
+    for _i in range(255):
+        y_prev = y
+        y = (y*y + c) / (2 * y + b - D)
+        # Equality with the precision of 1
+        if y > y_prev:
+            if y - y_prev <= 1:
+                return y
+        else:
+            if y_prev - y <= 1:
+                return y
+    raise
+
+
+@view
+@internal
+def _calc_withdraw_one_coin(_token_amount: uint256, i: int128, _use_underlying: bool) -> (uint256, uint256):
+    # First, need to calculate
+    # * Get current D
+    # * Solve Eqn against y_i for D - _token_amount
+    amp: uint256 = self._A()
+    rates: uint256[N_COINS] = self._stored_rates()
+    xp: uint256[N_COINS] = self._xp(rates)
+    D0: uint256 = self.get_D(xp, amp)
+
+    total_supply: uint256 = ERC20(self.lp_token).totalSupply()
+
+    D1: uint256 = D0 - _token_amount * D0 / total_supply
+    new_y: uint256 = self.get_y_D(amp, i, xp, D1)
+
+    xp_reduced: uint256[N_COINS] = xp
+    _fee: uint256 = self.fee * N_COINS / (4 * (N_COINS - 1))
+    rate: uint256 = rates[i]
+
+    for j in range(N_COINS):
+        dx_expected: uint256 = 0
+        xp_j: uint256 = xp[j]
+        if j == i:
+            dx_expected = xp_j * D1 / D0 - new_y
+        else:
+            dx_expected = xp_j - xp_j * D1 / D0
+        xp_reduced[j] -= _fee * dx_expected / FEE_DENOMINATOR
+
+    dy: uint256 = xp_reduced[i] - self.get_y_D(amp, i, xp_reduced, D1)
+    dy = (dy - 1) * PRECISION / rate  # Withdraw less to account for rounding errors
+    dy_fee: uint256 = ((xp[i] - new_y) * PRECISION / rate) - dy
+    if _use_underlying:
```

Comment:
Ouch. Shouldn't we convert `dy_fee` here, too?

Reply:
Or, really, like in 3pool, can have `dy` and `dy_0`, convert both and return `dy, dy - dy_0` or something

Reply:
This branch is only reached via `calc_withdraw_one_coin` where we're only returning dy, not the fee.

Took me a moment to remember why - this definitely deserves a comment.